File: java/FakeStatsServiceBuilder.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class FakeStatsServiceBuilder {
    private String defaultCustomerData = "19,123456.78;22,123456.78;57,123456.78";
    private Map<Integer,Double> builderData = null;

    public StatsService build() {
        String initialisationData = "";

        if (builderData == null) {
            initialisationData = defaultCustomerData;
        } else {
            String separator = "";

            for (Map.Entry<Integer, Double> entry : builderData.entrySet()) {
                initialisationData += separator + entry.getKey() + "," + entry.getValue();
                separator = ";";
            }
        }
        return new FakeStatsService(initialisationData);
    }

    public FakeStatsServiceBuilder withCustomerRevenue(int id, double revenue) {
        if (builderData == null) {
          builderData = new HashMap<Integer,Double>();
        }

        builderData.put(id, revenue);

        return this;
    }
}

File: cpp/fake_stats_service_builder.cpp
#include "fake_stats_service_builder.hpp"

#include <sstream>

namespace shouty_report
{
    stats_service fake_stats_service_builder::build() const
    {
        if (builder_data.size() == 0)
        {
            return fake_stats_service(default_customer_data);
        }
        else
        {
            std::ostringstream data;
            std::string separator = "";

            for (auto & entry : builder_data)
            {
                data << separator << entry.first << "," << entry.second;
                separator = ";";
            }

            return fake_stats_service(data.str());
        }
    }

    fake_stats_service_builder &
    fake_stats_service_builder::with_customer_revenue(
        int id,
        const std::string&  revenue
    )
    {
        builder_data[id] = revenue;

        return *this;
    }
}
